Libraries/VLM/Models/Paligemma.swift

```diff
@@ -0,0 +1,727 @@
+        PaliGemmaConfiguration.self,
+        from: try Data(contentsOf: url.appendingPathComponent("config.json")))
+    let model = PaliGemma(configuration)
+    return model
+}
+
+// TODO copied from LLM/Load -- ideally all of this moves into MLXCommon
+func prepareModelDirectory(
+    hub: HubApi, configuration: ModelConfiguration,
+    progressHandler: @Sendable @escaping (Progress) -> Void
+) async throws -> URL {
+    do {
+        switch configuration.id {
+        case .id(let id):
+            // download the model weights
+            let repo = Hub.Repo(id: id)
+            let modelFiles = ["*.safetensors", "*.json"]
+            return try await hub.snapshot(
+                from: repo, matching: modelFiles, progressHandler: progressHandler)
+
+        case .directory(let directory):
+            return directory
+        }
+    } catch Hub.HubClientError.authorizationRequired {
+        // an authorizationRequired means (typically) that the named repo doesn't exist on
+        // on the server so retry with local only configuration
+        return configuration.modelDirectory(hub: hub)
+    } catch {
+        let nserror = error as NSError
+        if nserror.domain == NSURLErrorDomain && nserror.code == NSURLErrorNotConnectedToInternet {
+            // Error Domain=NSURLErrorDomain Code=-1009 "The Internet connection appears to be offline."
+            // fall back to the local directory
+            return configuration.modelDirectory(hub: hub)
+        } else {
+            throw error
+        }
+    }
+}
+
+// TODO move to Configuration-ish place or MLXCommon/Load
+public func loadPaliGemma(hub: HubApi = HubApi()) async throws -> (
+    PaliGemma, Tokenizer, (CIImage) -> MLXArray
+) {
+    let configuration = ModelConfiguration(id: "mlx-community/paligemma-3b-mix-448-8bit")
+
+    let modelDirectory = try await prepareModelDirectory(
+        hub: hub, configuration: configuration, progressHandler: { _ in })
+    let model = try loadConfiguration(url: modelDirectory)
+    let tokenizer = try await loadTokenizer(configuration: configuration, hub: hub)
+
+    return (
+        model, tokenizer, PaliGemmaImageProvider(model.config.visionConfiguration).convert(image:)
+    )
+}
+
+// MARK: - Configuration
+
+public struct PaliGemmaConfiguration: Codable, Sendable {
+
+    public struct TextConfiguration: Codable, Sendable {
+        public let modelType: String
+        public let hiddenSize: Int
+        public let hiddenLayers: Int
+        public let intermediateSize: Int
+        public let attentionHeads: Int
+        public let kvHeads: Int
+        public let vocabularySize: Int
+        private let _rmsNormEps: Float?
+        public var rmsNormEps: Float { _rmsNormEps ?? 1e-6 }
+        private let _ropeTheta: Float?
+        public var ropeTheta: Float { _ropeTheta ?? 10_000 }
+        public let _ropeTraditional: Bool?
+        public var ropeTraditional: Bool { _ropeTraditional ?? false }
```

Comment:
Rather than doing the full implementation of Codable I went a simpler route for default values.